### DISCUSSION TITLE
Bootstrap with ethereal chroot method

### DISCUSSION THREAD

**Discussion by bwoodsend** [void-linux/void-packages#44842] [Q&A] [answered]
I'm trying to build packages for `aarch64` on an intel machine using Docker with Qemu architecture emulation. Qemu clashes with the default chroot method (gives `ERROR unshare (Invalid argument)`) so, since this is a throwaway container, I'm trying to use [ethereal](https://github.com/void-linux/void-packages/tree/fb5bd1a88bb2467c68388732a69a7c29b355abf7#ethereal).

To summarise my where I've got to, this is what I'm running:
```bash
git clone git@github.com:void-linux/void-packages.git
cd void-packages
docker run --rm -it --privileged -v "$PWD:/io" --platform=linux/aarch64 ghcr.io/void-linux/void-linux:latest-mini-aarch64-musl
```
Then inside the container:
```bash
xbps-install -ySu xbps git bash util-linux
bash
cd /io/
git config --global --add safe.directory /io
echo -e 'XBPS_CHROOT_CMD=ethereal\nXBPS_ALLOW_CHROOT_BREAKOUT=yes' > etc/conf
./xbps-src binary-bootstrap
./xbps-src pkg python3-ultrajson  # package chosen arbitrarily 
```
The `./xbps-src binary-bootstrap` step appears to do absolutely nothing – it exits immediately with exit code 0 and no output without populating `masterdir` then `./xbps-src pkg $anything` gives me `=> ERROR: Bootstrap not installed in /io/masterdir, can't continue.`. 

Is there something different I have to do instead of `binary-bootstrap` to bootstrap with `ethereal`?

**Comment by classabbyamp** [comment]
you may want to take a look at what our CI does, as it uses the ethereal method in a [ghcr.io/void-linux/xbps-src-masterdir](https://github.com/void-linux/void-docker/pkgs/container/xbps-src-masterdir) container

notably, see:
https://github.com/void-linux/void-packages/blob/a1618339163c0ffe72f116830b1fb13ae9bd698b/.github/workflows/build.yaml#L64-L90

https://github.com/void-linux/void-packages/blob/master/common/travis/prepare.sh
https://github.com/void-linux/void-packages/blob/master/common/travis/build.sh

**Reply by bwoodsend** [reply, marked as answer]
[This](https://github.com/void-linux/void-packages/blob/a1618339163c0ffe72f116830b1fb13ae9bd698b/common/travis/prepare.sh#L16C1-L16C18) is the line I was missing. Thanks!

**Comment by Johnnynator** [comment]
Your docker container IS the masterdir in the case of the ethereal method.

I would recommend skip the docker part if your host is running void (or at least provides xbps).
Creating and using non native masterdirs works fine as long as the foreign architecture is registered with binfmt_misc 

The way I would do it on void would be like this:
- Install `qemu-user-static` and `binfmt-support`
- enable the `binfmt-support` service
- `./xbps-src binary-bootstrap -m masterdir-aarch64 aarch64`
- `./xbps-pkg -m masterdir-aarch64 pkg <pkg>`
